language/move-prover/boogie-backend/src/spec_translator.rs

```diff
@@ -345,24 +345,32 @@ impl<'env> SpecTranslator<'env> {
 // ============================
 
 impl<'env> SpecTranslator<'env> {
-    pub(crate) fn finalize(&self) {
+    pub(crate) fn finalize(&mut self) {
         self.translate_choice_functions();
     }
 
     /// Translate lifted functions for choice expressions.
-    fn translate_choice_functions(&self) {
+    fn translate_choice_functions(&mut self) {
         let env = self.env;
         let infos_ref = self.lifted_choice_infos.borrow();
         // need the sorting here because `lifted_choice_infos` is a hashmap while we want
         // deterministic ordering of the output. Sorting uses the `.id` field, which represents the
         // insertion order.
-        let infos_sorted = infos_ref.values().sorted_by(|v1, v2| v1.id.cmp(&v2.id));
-        for info in infos_sorted {
+        let infos_sorted_with_keys = infos_ref.iter().sorted_by(|v1, v2| v1.1.id.cmp(&v2.1.id));
+        let orignal_spec_translator = self.clone();
+        for (key, info) in infos_sorted_with_keys {
             let fun_name = boogie_choice_fun_name(info.id);
             let result_ty = &env.get_node_type(info.node_id);
             let exp_loc = env.get_node_loc(info.node_id);
             let var_name = info.var.display(env.symbol_pool()).to_string();
             self.writer.set_location(&exp_loc);
+
+            if !key.1.is_empty() {
+                self.type_inst = key.1.clone();
+            } else {
+                self.type_inst = orignal_spec_translator.type_inst.clone();
```

Comment:
This should be unnecessary because it is always empty. You can add an assert! for this to be on the safe side.